### PR TITLE
Implement data sources for all variable types

### PR DESCRIPTION
**Pull request by timofurrer**
This implements a data source to retrieve a single variable and one to retrieve all variables for each variable type (project, group and instance level)

Closes: #386

### CODE REVIEW

File: internal/provider/data_source_gitlab_group_variable_test.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(`
+					resource "gitlab_group_variable" "this" {
+						group             = %d
+						key               = "any_key"
+					    value             = "any-value"
+						environment_scope = "*"
+					}
```
Comment (jtymes) [inline internal/provider/data_source_gitlab_group_variable_test.go:27]:
```suggestion
					resource "gitlab_group_variable" "this" {
						group             = %d
						key               = "any_key"
					        value             = "any-value"
						environment_scope = "*"
					}
```

File: internal/provider/data_source_gitlab_group_variables.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/xanzy/go-gitlab"
+)
+
+var _ = registerDataSource("gitlab_group_variables", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_group_variables`" + ` data source allows to retrieve all CI/CD variables in a group.
```
Comment (jtymes) [inline internal/provider/data_source_gitlab_group_variables.go:14]: I'd suggest using GitLab's term
```suggestion
		Description: `The ` + "`gitlab_group_variables`" + ` data source allows to retrieve all group-level CI/CD variables.
```

File: internal/provider/data_source_gitlab_instance_variable.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/xanzy/go-gitlab"
+)
+
+var _ = registerDataSource("gitlab_instance_variable", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_instance_variable`" + ` data source allows to retrieve details about a CI/CD variable registered on an instance level.
```
Comment (jtymes) [inline internal/provider/data_source_gitlab_instance_variable.go:13]:
```suggestion
		Description: `The ` + "`gitlab_instance_variable`" + ` data source allows to retrieve details about an instance-level CI/CD variable.
```

File: internal/provider/data_source_gitlab_instance_variable_test.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: `
+					resource "gitlab_instance_variable" "this" {
+						key               = "any_key"
+					    value             = "any-value"
+					}
```
Comment (jtymes) [inline internal/provider/data_source_gitlab_instance_variable_test.go:21]:
```suggestion
					resource "gitlab_instance_variable" "this" {
						key               = "any_key"
					        value             = "any-value"
					}
```

File: internal/provider/data_source_gitlab_instance_variables.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/xanzy/go-gitlab"
+)
+
+var _ = registerDataSource("gitlab_instance_variables", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_instance_variables`" + ` data source allows to retrieve all CI/CD variables regstered on an instance level.
```
Comment (jtymes) [inline internal/provider/data_source_gitlab_instance_variables.go:13]:
```suggestion
		Description: `The ` + "`gitlab_instance_variables`" + ` data source allows to retrieve all instance-level CI/CD variables.
```

File: internal/provider/data_source_gitlab_project_variable.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/xanzy/go-gitlab"
+)
+
+var _ = registerDataSource("gitlab_project_variable", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_project_variable`" + ` data source allows to retrieve details about a CI/CD variable in a project.
```
Comment (jtymes) [inline internal/provider/data_source_gitlab_project_variable.go:14]:
```suggestion
		Description: `The ` + "`gitlab_project_variable`" + ` data source allows to retrieve details about a project-level CI/CD variable.
```

File: internal/provider/data_source_gitlab_project_variable_test.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(`
+					resource "gitlab_project_variable" "this" {
+						project           = %d
+						key               = "any_key"
+					    value             = "any-value"
+						environment_scope = "*"
+					}
```
Comment (jtymes) [inline internal/provider/data_source_gitlab_project_variable_test.go:27]:
```suggestion
					resource "gitlab_project_variable" "this" {
						project           = %d
						key               = "any_key"
					        value             = "any-value"
						environment_scope = "*"
					}
```

File: internal/provider/data_source_gitlab_project_variables.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/xanzy/go-gitlab"
+)
+
+var _ = registerDataSource("gitlab_project_variables", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_project_variables`" + ` data source allows to retrieve all CI/CD variables in a project.
```
Comment (jtymes) [inline internal/provider/data_source_gitlab_project_variables.go:14]:
```suggestion
		Description: `The ` + "`gitlab_project_variables`" + ` data source allows to retrieve all project-level CI/CD variables.
```

File: internal/provider/resource_gitlab_group_variable.go [hunk 9]
```diff
@@ -25,53 +25,7 @@ var _ = registerResource("gitlab_group_variable", func() *schema.Resource {
 			StateContext: schema.ImportStatePassthroughContext,
 		},
```
Comment (jtymes) [inline internal/provider/resource_gitlab_group_variable.go:27]: I know it wasn't modified in this PR but I'd suggest changing the description of the resource to match the GitLab API docs like the other suggestions I've made:
```
Description: `The ` + "`" + `gitlab_group_variable` + "`" + ` resource allows to manage the lifecycle of a group-level CI/CD variable.
```

File: internal/provider/resource_gitlab_instance_variable.go [hunk 10]
```diff
@@ -24,40 +24,7 @@ var _ = registerResource("gitlab_instance_variable", func() *schema.Resource {
 			StateContext: schema.ImportStatePassthroughContext,
```
Comment (jtymes) [inline internal/provider/resource_gitlab_instance_variable.go:24]: I know it wasn't modified in this PR but I'd suggest changing the description of the resource to match the GitLab API docs like the other suggestions I've made:

```
Description: `The ` + "`" + `gitlab_instance_variable` + "`" + ` resource allows to manage the lifecycle of an instance-level CI/CD variable.
```

**Review by jtymes — CHANGES_REQUESTED** [review]
Functionally, everything looks great. Just a few minor suggestions.

File: internal/provider/resource_gitlab_instance_variable.go (hunk 10 above)
Comment (timofurrer) [reply internal/provider/resource_gitlab_instance_variable.go:24]: done :)

**Review by jtymes — APPROVED** [review]

**Review by armsnyder — APPROVED** [review]
Great improvements! 🚢

**Pull request merged**